File: goldencheck/llm/sample_block.py
"""Build representative sample blocks from DataFrame + findings."""
from __future__ import annotations
import logging
import random
from collections import defaultdict
import polars as pl
from goldencheck.models.finding import Finding

logger = logging.getLogger(__name__)


def build_sample_blocks(
    df: pl.DataFrame,
    findings: list[Finding],
    max_columns: int = 50,
    focus_columns: set[str] | None = None,
) -> dict[str, dict]:
    """Build a representative sample block for each column."""
    random.seed(42)

    # If focus_columns is specified, filter to only those columns early
    columns = list(df.columns)
    if focus_columns is not None:
        columns = [c for c in columns if c in focus_columns]

    # If too many columns, prioritize those with most findings
    if len(columns) > max_columns:
        logger.warning(
            "LLM boost limited to %d columns (dataset has %d). "
            "Columns with most findings prioritized.", max_columns, len(columns)
        )
        finding_counts = defaultdict(int)
        for f in findings:
            finding_counts[f.column] += 1
        columns = sorted(columns, key=lambda c: finding_counts[c], reverse=True)[:max_columns]

    # Index findings by column
    findings_by_col = defaultdict(list)
    for f in findings:
        findings_by_col[f.column].append(f)

    blocks = {}
    for col_name in columns:
        col = df[col_name]
        non_null = col.drop_nulls()

        # Metadata
        block: dict = {
            "column": col_name,
            "dtype": str(col.dtype),
            "row_count": len(df),
            "null_count": col.null_count(),
            "null_pct": round(col.null_count() / len(df), 3) if len(df) > 0 else 0,
            "unique_count": non_null.n_unique() if len(non_null) > 0 else 0,
        }

        # Top values (most frequent)
        if len(non_null) > 0:
            vc = non_null.value_counts().sort("count", descending=True)
            col_val_name = vc.columns[0]
            top = vc.head(5)
            block["top_values"] = [
                {"value": str(row[col_val_name]), "count": row["count"]}
                for row in top.iter_rows(named=True)
            ]

            # Rare values (least frequent)
            rare = vc.tail(5)
            block["rare_values"] = [
                {"value": str(row[col_val_name]), "count": row["count"]}
                for row in rare.iter_rows(named=True)
            ]

            # Random sample from middle
            all_vals = non_null.to_list()
            sample_size = min(5, len(all_vals))
            block["random_sample"] = [str(v) for v in random.sample(all_vals, sample_size)]
        else:
            block["top_values"] = []
            block["rare_values"] = []
            block["random_sample"] = []

        # Flagged values from profiler findings
        flagged = set()
        for f in findings_by_col.get(col_name, []):
            flagged.update(f.sample_values)
        block["flagged_values"] = list(flagged)

        # Existing findings
        block["existing_findings"] = [
            {"severity": f.severity.name.lower(), "check": f.check, "message": f.message}
            for f in findings_by_col.get(col_name, [])
        ]

        blocks[col_name] = block

    return blocks

File: packages/goldencheck-js/src/core/llm/sample-block.ts
/**
 * Build representative sample blocks from TabularData + findings.
 * Port of goldencheck/llm/sample_block.py.
 * Edge-safe: no Node.js dependencies.
 */

import type { TabularData } from "../data.js";
import type { Finding } from "../types.js";
import { createRng } from "../stats.js";

export interface ValueCount {
  value: string;
  count: number;
}

export interface ExistingFinding {
  severity: string;
  check: string;
  message: string;
}

export interface SampleBlock {
  column: string;
  dtype: string;
  row_count: number;
  null_count: number;
  null_pct: number;
  unique_count: number;
  top_values: ValueCount[];
  rare_values: ValueCount[];
  random_sample: string[];
  flagged_values: string[];
  existing_findings: ExistingFinding[];
}

/**
 * Build a representative sample block for each column.
 * Prioritizes columns with the most findings when exceeding maxColumns.
 */
export function buildSampleBlocks(
  data: TabularData,
  findings: readonly Finding[],
  maxColumns: number = 50,
  focusColumns?: ReadonlySet<string>,
): Record<string, SampleBlock> {
  // Deterministic RNG with seed=42
  const rng = createRng(42);

  // Filter columns
  let columns = [...data.columns];
  if (focusColumns) {
    columns = columns.filter((c) => focusColumns.has(c));
  }

  // If too many columns, prioritize those with most findings
  if (columns.length > maxColumns) {
    const findingCounts = new Map<string, number>();
    for (const f of findings) {
      findingCounts.set(f.column, (findingCounts.get(f.column) ?? 0) + 1);
    }
    columns.sort((a, b) => (findingCounts.get(b) ?? 0) - (findingCounts.get(a) ?? 0));
    columns = columns.slice(0, maxColumns);
  }

  // Index findings by column
  const findingsByCol = new Map<string, Finding[]>();
  for (const f of findings) {
    let arr = findingsByCol.get(f.column);
    if (!arr) {
      arr = [];
      findingsByCol.set(f.column, arr);
    }
    arr.push(f);
  }

  const blocks: Record<string, SampleBlock> = {};

  for (const colName of columns) {
    const nonNull = data.dropNulls(colName);
    const nullCount = data.nullCount(colName);
    const rowCount = data.rowCount;

    const block: SampleBlock = {
      column: colName,
      dtype: data.dtype(colName),
      row_count: rowCount,
      null_count: nullCount,
      null_pct: rowCount > 0 ? Math.round((nullCount / rowCount) * 1000) / 1000 : 0,
      unique_count: data.nUnique(colName),
      top_values: [],
      rare_values: [],
      random_sample: [],
      flagged_values: [],
      existing_findings: [],
    };

    if (nonNull.length > 0) {
      // Value counts sorted descending
      const vc = data.valueCounts(colName);
      const sorted = [...vc.entries()].sort((a, b) => b[1] - a[1]);

      // Top values (most frequent)
      block.top_values = sorted.slice(0, 5).map(([value, count]) => ({
        value: String(value),
        count,
      }));

      // Rare values (least frequent)
      block.rare_values = sorted.slice(-5).map(([value, count]) => ({
        value: String(value),
        count,
      }));

      // Random sample from non-null values (Fisher-Yates partial shuffle)
      const allVals = nonNull.map(String);
      const sampleSize = Math.min(5, allVals.length);
      const indices = Array.from({ length: allVals.length }, (_, i) => i);
      for (let i = 0; i < sampleSize; i++) {
        const j = i + Math.floor(rng() * (indices.length - i));
        [indices[i], indices[j]] = [indices[j]!, indices[i]!];
      }
      block.random_sample = indices.slice(0, sampleSize).map((idx) => allVals[idx]!);
    }

    // Flagged values from profiler findings
    const colFindings = findingsByCol.get(colName) ?? [];
    const flagged = new Set<string>();
    for (const f of colFindings) {
      for (const sv of f.sampleValues) {
        flagged.add(sv);
      }
    }
    block.flagged_values = [...flagged];

    // Existing findings summary
    block.existing_findings = colFindings.map((f) => ({
      severity: severityToLabel(f.severity),
      check: f.check,
      message: f.message,
    }));

    blocks[colName] = block;
  }

  return blocks;
}

// --- Helper ---

function severityToLabel(sev: number): string {
  switch (sev) {
    case 3:
      return "error";
    case 2:
      return "warning";
    case 1:
      return "info";
    default:
      return "warning";
  }
}
